python/entrance/ws_handler.py

```diff
@@ -57,15 +57,16 @@ async def handle_incoming_requests(self):
                 req = await self.ws.recv()
                 got_req = True
             except (asyncio.CancelledError, ConnectionClosed):
-                log.info("Websocket closed")
-                for feature in self.features:
-                    feature.close()
+                self._handle_websocket_closed()
                 break
             except Exception as e:
                 log.error("Websocket recv exception: %s", e)
             try:
                 if got_req:
                     await self._handle_req(req)
+            except (asyncio.CancelledError, ConnectionClosed):
+                self._handle_websocket_closed()
+                break
```

Comment:
Isn't it desirable to propagate `CancelledError` here?

https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError:
> In almost all situations the exception must be re-raised.

Reply:
If it is, this isn't done in the exception handling for ```await self.ws.recv()``` a couple lines above, and wasn't done thus far as all Exceptions were swallowed below.

Because of this I've largely kept the handling the same for ```asyncio.CancelledError```. Truth be told I think for our use case we only need to catch ```ConnectionClosed``` - I'm not sure when a ```CancelledError``` would be raised.

Reply:
So it looks from c3118826 that long enough ago, `asyncio.CancelledError` was raised on websocket close, and there's an argument for taking it out now and testing that nothing regresses.

But since this is the event loop for the websocket, it also doesn't seem unreasonable to catch it here (there's nowhere productive to propagate it to) so I think it's fine to leave in as-is. If someone is motivated to test without it, by all means go for it!